### ISSUE TITLE
Optimize the search with skipping root hidden folders.

### ISSUE THREAD

**Issue by NataliaBondarenko** [victordomingos/Count-files#93]
Hello! How about reducing the file checks? Please try this, I think it can be useful. 
Example, skipping root hidden folders for each loop: 
` if not include_hidden: if is_hidden_file_or_dir(root): continue`
in def search_files()
```
   for root, dirs, files in os.walk(dirpath):
            if not include_hidden:
                if is_hidden_file_or_dir(root):
                    print('SKIPPED:')
                    print('ROOT:', root, '- is hidden')
                    print(len(files), 'skipped files in hidden root')
                    # if root is hidden skip all files
                    continue
            for f in files:
                # do necessary checks with files from not hidden folders
```
Example output:
![output](https://user-images.githubusercontent.com/23127253/45931917-c1eaae00-bf7d-11e8-9918-72ecc7d1d4f3.png)

Thus, you can skip unnecessary scanning of files/dirs located in known hidden folders.
I think this will allow us to speed up the processing of files a little.
Also, if the root does not contain a hidden folders, then you can only check the file names for the presence of a point (for Unix). For Windows it's more difficult, since attributes are requested for each part of the path.

**Comment by NataliaBondarenko** [comment]
Then we can do this:
For Windows
in `def is_hidden_file_or_dir`
```
...
    platform_name = sys.platform
    if platform_name.startswith('win'):
        if os.path.isfile(filepath):
            # it is filepath, if root is checked - no need to make full list again
            list_for_check = [filepath]
        else:
            # it is dirpath
            # list with full paths of all parents in filepath except drive
            list_for_check = list(Path(filepath).parents)[:-1]
            list_for_check.append(Path(filepath))
        response = []
        for some_path in list_for_check:
            try:
                attrs = ctypes.windll.kernel32.GetFileAttributesW(str(some_path))
                assert attrs != -1
                result = bool(attrs & 2)
...
```

**Comment by victordomingos** [comment]
Please feel free to submit a PR for this, in case it results in a mensurable performance advantage.

**Comment by NataliaBondarenko** [comment]
I did some tests.
For example, for the query: `count-files ~\Count-files -fe .`
- Option with skipping hidden root folders
timeit test:
0.44813516490784727   <-first run
0.5181522931553151
0.29632153210006906
0.26906287275605134
0.26137796864914914
0.3056913220680705
and so on
cProfile test:
63129 function calls (59741 primitive calls) in 0.280 seconds
- Current option
timeit test:
12.263768588613827   <-first run
1.7576878931754472
1.713331801447005
1.4804995241839427
1.3633438607754804
1.3834991894876865
and so on
cProfile test:
401598 function calls (383635 primitive calls) in 1.696 seconds

**Comment by NataliaBondarenko** [comment]
It is necessary to add to every function that processes files
```
if not include_hidden:
    if is_hidden_file_or_dir(root):
        continue
``` 
after `for root, dirs, files in os.walk(dirpath):`.
If I don’t figure out how to organize it better, then I’ll do it.

**Comment by victordomingos** [comment]
It’s a nice speed improvement. ;) Please make sure it does not have any side effects.

**Comment by NataliaBondarenko** [comment]
To speed up the program on Windows, we need to change the structure. In this case, if we add a root folder check, it will be more beneficial for Windows than for Linux or Mac OS.

**Comment by victordomingos** [comment]
> To speed up the program on Windows, we need to change the structure. In this case, if we add a root folder check, it will be more beneficial for Windows than for Linux or Mac OS.

As long as the code doesn't become harder to read and as long as there are no issues on other platforms, I think that improving speed is always a good thing. ;)

**Comment by victordomingos** [comment]
Since we are using the same "hidden" definition on both Linux and macOS, the results should be roughly the same on Mac. There are a other subtleties to hidden files and folders on macOS, that we may take a look in the future, and that could eventually bring us to results more similar to those of Windows. But for now, there is one special kind of volume that is not being ignored and should be, when ignoring hidden files and folders, in my opinion. Some info on this page:

https://apple.stackexchange.com/questions/18846/mtmfs-consuming-an-entire-cpu-core-and-fans-running-full-bore-with-lion

Basically, recent macOS versions have a local backup feature that uses a special file system that from time to time is mounted on `/Volumes/MobileBackups`. I had that happening on my Mac once, and I noticed that count-files was taking much longer than usual.

**Comment by NataliaBondarenko** [comment]
Hello! Preliminary test results for Windows.
recursive=True, include_hidden=False

<table>
  <tr>
    <th>action</th>
    <th> many hidden files and folders</th>
    <th>no hidden files or folders</th>
  </tr>
<tr>
    <td>count</td>
   <td>faster</td>
   <td>faster</td>
  </tr>
<tr>
    <td>search <b>-fe ..</b></td>
   <td>faster</td>
   <td>faster</td>
  </tr>
<tr>
   <td>total <b>-t ..</b></td>
   <td>faster</td>
   <td>faster</td>
  </tr>
  <tr>
    <td>search <b>-fe ext</b></td>
    <td colspan="2">depends on the number of files and whether they are hidden and located in hidden </td>
  </tr>
  <tr>
    <td>total <b>-t ext</b></td>
    <td colspan="2">depends on the number of files and whether they are hidden and located in hidden </td>
  </tr>
</table>

For count, search `-fe ..` and total `-t ..`:
In this case, we know in advance that we need to check absolutely all files. 
- file processing is two or more times faster if there are a lot of hidden files and folders
- if there are no hidden folders and files, then it is not much faster

Note: if the number of files is less than folders, then slower (depending on the difference in the number)
For example, if there are many empty subfolders: `/folder/subfolder/subfolder/one_file_in_a_folder`.
But I think this is a rare case.

For search `-fe ext` and total `-t ext`:
It depends on the situation that we do not know in advance. In this case, optimization is impractical.

Thus, we can make partial changes (for count, search `-fe ..` and total `-t ..`).
